File: lincheck/src/main/java/org/jetbrains/kotlinx/lincheck/verifier/quiescent/QuiescentConsistencyVerifier.java
package org.jetbrains.kotlinx.lincheck.verifier.quiescent;

/*
 * #%L
 * Lincheck
 * %%
 * Copyright (C) 2015 - 2018 Devexperts, LLC
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Lesser Public License for more details.
 * 
 * You should have received a copy of the GNU General Lesser Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/lgpl-3.0.html>.
 * #L%
 */

import org.jetbrains.kotlinx.lincheck.Actor;
import org.jetbrains.kotlinx.lincheck.execution.ExecutionResult;
import org.jetbrains.kotlinx.lincheck.execution.ExecutionScenario;
import org.jetbrains.kotlinx.lincheck.verifier.CachedVerifier;
import org.jetbrains.kotlinx.lincheck.verifier.linearizability.LinearizabilityVerifier;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class QuiescentConsistencyVerifier extends CachedVerifier {
    private final ExecutionScenario originalScenario;
    private final LinearizabilityVerifier linearizabilityVerifier;

    public QuiescentConsistencyVerifier(ExecutionScenario scenario, Class<?> testClass) {
        this.originalScenario = scenario;
        this.linearizabilityVerifier = new LinearizabilityVerifier(convertScenario(scenario), testClass);
    }

    private static ExecutionScenario convertScenario(ExecutionScenario scenario) {
        List<List<Actor>> newParallelExecution = convertAccordingToScenario(scenario.parallelExecution, scenario.parallelExecution);
        return new ExecutionScenario(scenario.initExecution, newParallelExecution, scenario.postExecution);
    }

    private static <T> List<List<T>> convertAccordingToScenario(List<List<Actor>> parallelExecution, List<List<T>> toConvert) {
        List<List<T>> res = new ArrayList<>();
        for (int t = 0; t < parallelExecution.size(); t++)
            res.add(new ArrayList<>());
        for (int t = 0; t < parallelExecution.size(); t++) {
            int nActors = parallelExecution.get(t).size();
            for (int i = 0; i < nActors; i++) {
                Actor actor = parallelExecution.get(t).get(i);
                T val = toConvert.get(t).get(i);
                if (isQuiescentConsistent(actor) && nActors > 1)
                    res.add(Collections.singletonList(val));
                else
                    res.get(t).add(val);
            }
        }
        return res;
    }

    private static boolean isQuiescentConsistent(Actor actor) {
        return actor.method.isAnnotationPresent(QuiescentConsistent.class);
    }

    @Override
    public boolean verifyResultsImpl(ExecutionResult results) {
        return linearizabilityVerifier.verifyResultsImpl(new ExecutionResult(
            results.initResults,
            convertAccordingToScenario(originalScenario.parallelExecution, results.parallelResults),
            results.postResults)
        );
    }
}

File: lincheck/src/main/java/org/jetbrains/kotlinx/lincheck/verifier/quiescent/QuiescentConsistencyVerifier.kt
/*-
 * #%L
 * Lincheck
 * %%
 * Copyright (C) 2019 - 2020 JetBrains s.r.o.
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Lesser Public License for more details.
 * 
 * You should have received a copy of the GNU General Lesser Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/lgpl-3.0.html>.
 * #L%
 */
package org.jetbrains.kotlinx.lincheck.verifier.quiescent

import org.jetbrains.kotlinx.lincheck.*
import org.jetbrains.kotlinx.lincheck.execution.*
import org.jetbrains.kotlinx.lincheck.verifier.*
import org.jetbrains.kotlinx.lincheck.verifier.linearizability.*

/**
 * This verifier tests for quiescent consistency.
 * Note that it does not support quiescent points.
 * Thus, it potentially does not find some bugs.
 * However, we believe that quiescent points do not occur
 * in practice while supporting them complicates the implementation.
 */
class QuiescentConsistencyVerifier(sequentialSpecification: Class<*>) : CachedVerifier() {
    private val linearizabilityVerifier = LinearizabilityVerifier(sequentialSpecification)
    override fun checkStateEquivalenceImplementation() = linearizabilityVerifier.checkStateEquivalenceImplementation()

    override fun verifyResultsImpl(scenario: ExecutionScenario, results: ExecutionResult): Boolean {
        val newThreads = scenario.threads + scenario.parallelExecution.flatten().count { it.isQuiescentConsistent }
        val parallelExecution = ArrayList<MutableList<Actor>>()
        val parallelResults = ArrayList<MutableList<ResultWithClock>>()
        repeat(scenario.threads) {
            parallelExecution.add(ArrayList())
            parallelResults.add(ArrayList())
        }
        val clocks = Array(scenario.threads) { ArrayList<IntArray>() }
        val clockMapping = Array(scenario.threads) { ArrayList<Int>() }
        clockMapping.forEach { it.add(-1) }
        scenario.parallelExecution.forEachIndexed { t, threadActors ->
            threadActors.forEachIndexed { i, a ->
                val r = results.parallelResultsWithClock[t][i]
                if (a.isQuiescentConsistent) {
                    clockMapping[t].add(clockMapping[t][i])
                    parallelExecution.add(mutableListOf(a))
                    parallelResults.add(mutableListOf(r.result.withEmptyClock(newThreads)))
                } else {
                    clockMapping[t].add(clockMapping[t][i] + 1)
                    parallelExecution[t].add(a)
                    val c = IntArray(newThreads) { 0 }
                    clocks[t].add(c)
                    parallelResults[t].add(ResultWithClock(r.result, HBClock(c)))
                }
            }
        }
        clocks.forEachIndexed { t, threadClocks ->
            threadClocks.forEachIndexed { i, c ->
                for (j in 0 until scenario.threads) {
                    val old = results.parallelResultsWithClock[t][i].clockOnStart[j]
                    c[j] = clockMapping[j][old]
                }
            }
        }
        val convertedScenario = ExecutionScenario(scenario.initExecution, parallelExecution, scenario.postExecution)
        val convertedResults = ExecutionResult(results.initResults, parallelResults, results.postResults)
        return linearizabilityVerifier.verifyResultsImpl(convertedScenario, convertedResults)
    }
}

private val Actor.isQuiescentConsistent: Boolean get() = method.isAnnotationPresent(QuiescentConsistent::class.java)

/**
 * This annotation indicates that the method it is presented on
 * is quiescent consistent.
 *
 * @see QuiescentConsistencyVerifier
 */
@Retention(AnnotationRetention.RUNTIME)
@Target(AnnotationTarget.FUNCTION, AnnotationTarget.PROPERTY_GETTER, AnnotationTarget.PROPERTY_SETTER)
annotation class QuiescentConsistent
